module/VuFind/src/VuFind/AjaxHandler/GetSearchResults.php

```diff
@@ -0,0 +1,389 @@
+     */
+    protected $elements = [
+        '.js-record-list' => [
+            'method' => 'renderResults',
+            'target' => 'outer',
+        ],
+        '.js-pagination' => [
+            'method' => 'renderPagination',
+            'target' => 'outer',
+        ],
+        '.js-pagination-simple' => [
+            'method' => 'renderPaginationSimple',
+            'target' => 'outer',
+        ],
+        '.js-search-stats' => [
+            'method' => 'renderSearchStats',
+            'target' => 'inner',
+            'attrs' => [
+                'aria-live' => 'polite',
+            ],
+        ],
+    ];
+
+    /**
+     * Constructor
+     *
+     * @param SessionSettings  $sessionSettings Session settings
+     * @param ResultsManager   $resultsManager  Results Manager
+     * @param PhpRenderer      $renderer        View renderer
+     * @param RecordLoader     $recordLoader    Record loader
+     * @param ?UserRow         $user            Logged-in user
+     * @param string           $sessionId       Session ID
+     * @param SearchNormalizer $normalizer      Search normalizer
+     * @param SearchTable      $searchTable     Search table
+     * @param array            $config          Main configuration
+     */
+    public function __construct(
+        SessionSettings $sessionSettings,
+        ResultsManager $resultsManager,
+        PhpRenderer $renderer,
+        RecordLoader $recordLoader,
+        ?UserRow $user,
+        string $sessionId,
+        SearchNormalizer $normalizer,
+        Search $searchTable,
+        array $config
+    ) {
+        $this->sessionSettings = $sessionSettings;
+        $this->resultsManager = $resultsManager;
+        $this->renderer = $renderer;
+        $this->recordLoader = $recordLoader;
+        $this->user = $user;
+        $this->sessionId = $sessionId;
+        $this->searchNormalizer = $normalizer;
+        $this->searchTable = $searchTable;
+        $this->config = $config;
+    }
+
+    /**
+     * Handle a request.
+     *
+     * @param Params $params Parameter helper from controller
+     *
+     * @return array [response data, HTTP status code]
+     */
+    public function handleRequest(Params $params)
+    {
+        $results = $this->getSearchResults($params);
+        if (!$results) {
+            return $this->formatResponse(['error' => 'Invalid request'], 400);
+        }
+        $elements = $this->getElements($params, $results);
+        return $this->formatResponse(compact('elements'));
+    }
+
+    /**
+     * Get search results
+     *
+     * @param Params $params Request params
+     *
+     * @return ?Results
+     */
+    protected function getSearchResults(Params $params): Results
+    {
+        parse_str($params->fromQuery('querystring', ''), $searchParams);
+
+        $backend = $params->fromQuery('source', DEFAULT_SEARCH_BACKEND);
+        $searchType = $params->fromQuery('searchType', '');
+        if ('versions' === $searchType) {
```

Comment:
I extracted a tiny bit to a new trait, but it's not much. The special case is needed because versions search doesn't have a lookfor parameter, to ensure that the work keys are current and to make sure it's not saved in history.

It would be neat if we could do most of this in initFromRequest, but then the Params object would need a record loader, and that feels wrong too.

Reply:
What if, instead of a trait, we had a `\VuFind\Record\VersionsHelper` service, which could in turn have a dependency on the RecordLoader? Then we could fetch the VersionsHelper instead of the RecordLoader in the RecordVersionsSearchTrait, and we could make the VersionsHelper a direct dependency of the AJAX handler instead of the RecordLoader. I think this would allow further consolidation of logic (e.g. retrieval of keys).

Reply:
An excellent idea, thanks! I was so concentrated on the trait approach that I couldn't think that far. :)

Reply:
No problem! Traits are very inviting because they're so easy, but I think they're actually only really good in a limited set of situations because they are hard to override/customize. I'm sure I haven't perfectly lived up to it, and I'm sure there are valid exceptions to the rule, but I'm trying to make this my rule of thumb: if it's behavior shared by a whole bunch of different things (i.e. a reusable pattern), it should be a trait. If it's centralized logic (i.e. complex behavior that does exactly the same thing in different parts of the code), it should be a service so that it can be overridden/extended.

Reply:
The change has been done, but I'll hold off any review requests until after #2876 and #2878 are done.